src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+            /// </summary>
+            private QubitTimeNode Sample;
+
+            /// <summary>
+            /// Resets minimum and maximum value of nodes seen from the set.
+            /// Also sets sample object, which should not be counted because it is not an object from the set.
+            /// </summary>
+            /// <param name="sample">Sample object that is not considered for minimum and maximum.</param>
+            internal void ResetForComparison(QubitTimeNode sample) {
+                MaxLowerBound = ComplexTime.MinValue;
+                MinUpperBound = ComplexTime.MaxValue;
+                Sample = sample;
+            }
+
+            /// <summary>
+            /// Compares Time field of two nodes. Also updates MaxLowerBound and MinUpperBound.
+            /// </summary>
+            /// <returns>Result of comparison of Time field of QubitTimeNode</returns>
+            public int Compare(QubitTimeNode a, QubitTimeNode b) {
+                // Note that comparison of a and b nodes to Sample is "by reference" in this function.
+                int result = ComplexTime.Compare(a.Time, b.Time);
+                if (result > 0) {
+                    if (a == Sample && b != Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, a.Time);
+                    }
+                } else if (result < 0) {
+                    if (a == Sample && b != Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, a.Time);
+                    }
+                } else {
+                    // We found the value if one argument is the sample and nother is not.
+                    if ((a == Sample) != (b == Sample)) {
+                        MaxLowerBound = b.Time;
+                        MinUpperBound = b.Time;
+                    }
+                }
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// QubitTimeNodes sorted by Time field. SortedSet won't accept values with the same Time field.
+        /// When QubitPool needs to store multiple qubits with the same Time, additional QubitTimeNode instances
+        /// are linked to the head node from this set.
+        /// </summary>
+        private readonly SortedSet<QubitTimeNode> QubitsSortedByTime;
+
+        /// <summary>
+        /// Comparer of two nodes, which compares them by Time field and collects additional data.
+        /// </summary>
+        private readonly VisitingComparer NodeComparer;
+
+        /// <summary>
+        /// Sample which is been compared to the nodes stored in the set.
+        /// Multiple comparisons are done by SortedSet to find if this Sample is present in the set.
```

Comment:
Is this a statement about how SortedSet is implemented in C# CLR? Do you rely on this particular implementation in some way?

Reply:
This is my second concern about this approach. Does this depend on CLR SortedSet implementation? Yes and No. To some extent. For any reasonable implementation should work. Of course one can specifically craft an implementation that would break this. For example, you can cache previously passed values and then compare against them (mostly for fun). I assume implementation is reasonable.